src/contexts/AuthProvider.js

```diff
@@ -0,0 +1,124 @@
+import { createContext, useContext, useEffect, useState } from "react";
+import jwt_decode from "jwt-decode";
+import * as authApi from "../api/auth";
+import * as userApi from "../api/user";
+import {
+  setupAuthExceptionHandler,
+  setupAuthHeader,
+  setupCancelToken,
+} from "../utils/helper";
+import { useNavigate } from "react-router";
+import showToast from "../utils/showToast";
+import axios from "axios";
+
+const AuthContext = createContext();
+
+const AuthProvider = ({ children }) => {
+  const token = JSON.parse(localStorage?.getItem("__auth_token")) || null;
+  const user = JSON.parse(localStorage?.getItem("__auth_user")) || null;
+  const [isUserLoading, setUserLoading] = useState(false);
+  const [tokenExpiry, setTokenExpiry] = useState(null);
+  const navigate = useNavigate();
+  const source = axios.CancelToken.source();
+  setupCancelToken(source);
+  token && setupAuthHeader(token);
+
+  useEffect(() => {
+    setupAuthExceptionHandler(signout, navigate);
+  }, []);
+
+  useEffect(() => {
+    if (token) {
+      try {
+        setUserLoading(true);
+        const decodedToken = jwt_decode(token);
+        setTokenExpiry(decodedToken.exp);
+        if (decodedToken.exp < Date.now() / 1000) {
+          throw new Error("Session Expired");
+        }
+      } catch (err) {
+        alert(`${err.message}\nPlease sign-in again`);
+        signout();
+        navigate("/signin");
```

Comment:
Can we auto-refresh the token?

It's not the user's fault that `Session Expired` then why ask him to log-in again?